wolfcrypt/src/asn.c

```diff
@@ -14394,28 +14421,38 @@ static int ConfirmNameConstraints(Signer* signer, DecodedCert* cert)
             case ASN_RFC822_TYPE:
                 /* Shouldn't it validade E= in subject as well? */
                 name = cert->altEmailNames;
-                break;
-            case ASN_DIR_TYPE:
-                if (cert->subjectRaw != NULL) {
+
+                /* Add subject email for checking. */
+                if (cert->subjectEmail != NULL) {
+                    /* RFC 5280 section 4.2.1.10
+                     * "When constraints are imposed on the rfc822Name name
+                     * form, but the certificate does not include a subject
+                     * alternative name, the rfc822Name constraint MUST be
+                     * applied to the attribute of type emailAddress in the
+                     * subject distinguished name" */
                     subjectDnsName.next = NULL;
-                    subjectDnsName.type = ASN_DIR_TYPE;
-                    subjectDnsName.len = cert->subjectRawLen;
-                    subjectDnsName.name = (char *)cert->subjectRaw;
-                    name = &subjectDnsName;
+                    subjectDnsName.type = ASN_RFC822_TYPE;
+                    subjectDnsName.len  = cert->subjectEmailLen;
+                    subjectDnsName.name = (char *)cert->subjectEmail;
                 }
+                break;
+            case ASN_DIR_TYPE:
+                name = cert->altDirNames;
 
-                #ifndef WOLFSSL_NO_ASN_STRICT
+            #ifndef WOLFSSL_NO_ASN_STRICT
                 /* RFC 5280 section 4.2.1.10
                     "Restrictions of the form directoryName MUST be
                     applied to the subject field .... and to any names
                     of type directoryName in the subjectAltName
                     extension"
                 */
-                if (name != NULL)
```

Comment:
Why not prepend?

name = cert->altDirNames
...
subjectDnsName.next = name;
name = subjectDnsName;

Reply:
I like prepending, it's more efficient (goes through the 'base' list only once). The change away from prepend was done to be more explicit and match the pattern used with both the email and DIR alt names. If prepending is ok I'd like that better, will update the PR.